edx/edx-enterprise
# ENT-202; ENT-205: Refactor _enroll_users; prevent duplicate messaging

**Description:**

This pull request removes the technical debt that has slowly sprung up around the `_enroll_users` method by refactoring it out into separate, smaller, methods (ENT-205). This allows us to provide better messaging and remove the duplicate messaging that was often caused as a result of enrolling pending learners in programs (ENT-202).

**JIRA:** [ENT-202](https://openedx.atlassian.net/browse/ENT-202); [ENT-205](https://openedx.atlassian.net/browse/ENT-205)

**Dependencies:** None

**Merge deadline:** 13 March 2017

**Installation instructions:**
Testing requires the following:
* Multiple courses (sharing at least one enrollment mode)
* A properly-configured Course Discovery API integration, populated with course metadata
* At least one program containing multiple courses set up in the Course Discovery API
* An EnterpriseCustomer

**Testing instructions:**

* Test for duplicate messages
    1. Open the EnterpriseCustomer's Manage Learners admin panel
    2. Enter an email address that does not exist in the system in the email box, and enter your program's UUID in the Program box. Select a course mode shared between all courses in the program, and click Submit.
    3. Observe that no duplicate messages are presented, but one message indicating the user was successfully linked to the EnterpriseCustomer, and one message indicating that a pending link was created for the program as a whole.
* Test refactoring
    1. Attempt the various actions permitted by the Manage Learners form, observing that behavior is identical to the behavior produced by the existing code, with some small exceptions around the messages that are rendered onscreen. Examples:
        * Check that an existing user can be linked and enrolled in a course, and that logging in with that user shows the course they were enrolled in.
        * Check that a non-existing user can be linked and enrolled in a course, and that registering with the email address used to link works and shows the course the pending enrollment was created for.
        * Using a valid email address you control (on a sandbox) or by observing the terminal output (on a devstack), verify that notification emails are sent for both program and course enrollments.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**
